packages/near-membrane-base/src/blue.ts

```diff
@@ -161,52 +152,92 @@ export function blueProxyFactory(env: MembraneBroker) {
-            if (typeof get === 'function') {
-                blueDescriptor.get = getBlueFunction(get);
+            if (get !== undefined) {
+                blueDescriptor.get = getBlueValue(get);
             }
-            if (typeof set === 'function') {
-                blueDescriptor.set = getBlueFunction(set);
+            if (set !== undefined) {
+                blueDescriptor.set = getBlueValue(set);
             }
         }
         return blueDescriptor;
     }
 
     function getBlueFunction(redFn: RedFunction): BlueFunction {
-        const blueFn = env.getBlueRef(redFn);
-        if (blueFn === undefined) {
-            return createBlueProxy(redFn) as BlueFunction;
+        // caching logic
+        const blueFn = WeakMapGet(env.redMap, redFn) as RedFunction | undefined;
+        if (blueFn !== undefined) {
+            return blueFn;
+        }
+        // extracting the metadata about the proxy target
+        let targetTraits = TargetTraits.IsFunction;
+        let targetFunctionName: string | undefined;
+        // detecting arrow function vs function
+        try {
+            targetTraits |= (!('prototype' in redFn) ? 1 : 0) << 3; // IsArrowFunction
+        } catch {
+            // target is either a revoked proxy, or a proxy that throws on the
+            // `has` trap, in which case going with a strict mode function seems
+            // appropriate.
         }
-        return blueFn as RedFunction;
+        try {
+            // a revoked proxy will throw when reading the function name
+            targetFunctionName = ReflectGetOwnPropertyDescriptor(redFn, 'name')?.value;
+        } catch {
+            // intentionally swallowing the error because this method is just extracting the
+            // function in a way that it should always succeed except for the cases in which
+            // the provider is a proxy that is either revoked or has some logic to prevent
+            // reading the name property descriptor.
+        }
+        return createBlueProxy(redFn, targetTraits, targetFunctionName) as BlueFunction;
+    }
+
+    function getBlueArrayOrObject(red: RedArrayOrObject): BlueArrayOrObject {
+        // caching logic
+        const blue: RedArrayOrObject | undefined = WeakMapGet(env.redMap, red);
+        if (blue !== undefined) {
+            return blue;
+        }
+        // extracting the metadata about the proxy target
+        let targetTraits = 0;
+        const targetFunctionName = undefined;
+        let targetIsArray = false;
+        try {
+            // try/catch in case Array.isArray throws when target is a revoked proxy
+            targetIsArray = isArrayOrNotOrThrowForRevoked(red);
+        } catch {
+            // target is a revoked proxy, so the type doesn't matter much from this point on
+        }
+
+        targetTraits |= (targetIsArray ? 1 : 0) << 0; // IsArray
+        targetTraits |= (targetIsArray ? 0 : 1) << 2; // IsObject
```

Comment:
Since `targetIsArray` will be a boolean, these can be written: 

```js
targetTraits |= +targetIsArray && TargetTraits.IsArray;
targetTraits |= +!targetIsArray && TargetTraits.IsObject;
```